File: ledger-models-javascript/node/wrappers/models/security/Issuance.ts
import { IssuanceProto } from "../../../fintekkers/models/security/bond/issuance_pb";
import { AuctionTypeProto } from "../../../fintekkers/models/security/bond/auction_type_pb";
import { DecimalValueProto } from "../../../fintekkers/models/util/decimal_value_pb";
import { LocalDateProto } from "../../../fintekkers/models/util/local_date_pb";
import { LocalDate } from "../utils/date";
import Decimal from "decimal.js";

/**
 * Typed wrapper around a single IssuanceProto. Returns null for unset
 * sub-messages and Decimal/LocalDate for populated ones — callers no longer
 * have to spell out the proto / Decimal coercions at every call site.
 *
 * Note: IssuanceProto has no `dated_date` or `auction_date` field on the
 * proto today, so those accessors are intentionally absent.
 */
class Issuance {
  proto: IssuanceProto;

  constructor(proto: IssuanceProto) {
    this.proto = proto;
  }

  private static _toDecimal(value: DecimalValueProto | undefined): Decimal | null {
    if (!value) return null;
    return new Decimal(value.getArbitraryPrecisionValue());
  }

  private static _toLocalDate(value: LocalDateProto | undefined): LocalDate | null {
    if (!value) return null;
    return new LocalDate(value);
  }

  /** Auction issue (settlement) date. Null if unset. */
  getIssueDate(): LocalDate | null {
    return Issuance._toLocalDate(this.proto.getAuctionIssueDate());
  }

  /** Auction announcement date. Null if unset. */
  getAnnouncementDate(): LocalDate | null {
    return Issuance._toLocalDate(this.proto.getAuctionAnnouncementDate());
  }

  /** Auction offering amount (original face value offered). Null if unset. */
  getOriginalFaceValue(): Decimal | null {
    return Issuance._toDecimal(this.proto.getAuctionOfferingAmount());
  }

  /** Total quantity accepted at auction. Null if unset. */
  getTotalAccepted(): Decimal | null {
    return Issuance._toDecimal(this.proto.getTotalAccepted());
  }

  /** Outstanding quantity after auction settles. Null if unset. */
  getPostAuctionOutstandingQuantity(): Decimal | null {
    return Issuance._toDecimal(this.proto.getPostAuctionOutstandingQuantity());
  }

  /** Quantity of the mature security used in reopenings. Null if unset. */
  getMatureSecurityAmount(): Decimal | null {
    return Issuance._toDecimal(this.proto.getMatureSecurityAmount());
  }

  /** Single-price auction clearing price. Null if unset (e.g., multi-price). */
  getPriceForSinglePriceAuction(): Decimal | null {
    return Issuance._toDecimal(this.proto.getPriceForSinglePriceAuction());
  }

  /** Auction type enum (defaults to UNKNOWN_AUCTION_TYPE if unset). */
  getAuctionType(): AuctionTypeProto {
    return this.proto.getAuctionType();
  }
}

export default Issuance;

File: ledger-models-java/src/main/java/common/models/security/bonds/Issuance.java
package common.models.security.bonds;

import fintekkers.models.security.bond.AuctionTypeProto;
import fintekkers.models.security.bond.IssuanceProto;
import protos.serializers.util.proto.ProtoSerializationUtil;

import java.math.BigDecimal;
import java.time.LocalDate;

/**
 * Typed wrapper over {@link IssuanceProto}. Exposes per-auction issuance
 * details (announcement/issue dates, offering amount, total accepted, etc.)
 * as Java domain types. Created when the parent BondSecurity unwraps its
 * stashed {@code SecurityProto.bond_details.issuance_info} list.
 *
 * <p>Accessors return {@code null} when the underlying proto field is unset.
 */
public class Issuance {
    private final IssuanceProto proto;

    public Issuance(IssuanceProto proto) {
        this.proto = proto;
    }

    public LocalDate getIssueDate() {
        if (!proto.hasAuctionIssueDate()) return null;
        return ProtoSerializationUtil.deserializeLocalDate(proto.getAuctionIssueDate());
    }

    public LocalDate getAnnouncementDate() {
        if (!proto.hasAuctionAnnouncementDate()) return null;
        return ProtoSerializationUtil.deserializeLocalDate(proto.getAuctionAnnouncementDate());
    }

    public BigDecimal getOriginalFaceValue() {
        if (!proto.hasAuctionOfferingAmount()) return null;
        return ProtoSerializationUtil.deserializeBigDecimal(proto.getAuctionOfferingAmount());
    }

    public BigDecimal getTotalAccepted() {
        if (!proto.hasTotalAccepted()) return null;
        return ProtoSerializationUtil.deserializeBigDecimal(proto.getTotalAccepted());
    }

    public BigDecimal getPostAuctionOutstandingQuantity() {
        if (!proto.hasPostAuctionOutstandingQuantity()) return null;
        return ProtoSerializationUtil.deserializeBigDecimal(proto.getPostAuctionOutstandingQuantity());
    }

    public BigDecimal getMatureSecurityAmount() {
        if (!proto.hasMatureSecurityAmount()) return null;
        return ProtoSerializationUtil.deserializeBigDecimal(proto.getMatureSecurityAmount());
    }

    public BigDecimal getPriceForSinglePriceAuction() {
        if (!proto.hasPriceForSinglePriceAuction()) return null;
        return ProtoSerializationUtil.deserializeBigDecimal(proto.getPriceForSinglePriceAuction());
    }

    public AuctionTypeProto getAuctionType() {
        return proto.getAuctionType();
    }

    /**
     * Raw escape hatch — primarily for serialization round-trips.
     * Prefer the typed accessors above.
     */
    IssuanceProto getProto() {
        return proto;
    }
}
